cilium: transparent encryption, use correct keys during update from old key to new key

jrfastab opened the pull request:
When updating a key in a shared secret not all nodes will receive the update "at the same time" so we need to handle case where keys are out of sync. Currently, we lookup which key to use in the ipcache which works great if the peer node is using no-key (null key 0) or has a key we've installed in the datapath already but if its a new key not in the datapath then trying to encrypt using the ID fails and traffic is dropped. To fix this keep the state of the last installed key and if the peers node is different use the previous key until the nodes sync at which point after the sync the current state will equal the peers key. This way we (a) always use keys installed on the node and (b) always use keys installed on the peer avoiding traffic drops in the intermediate state while keys sync and/or rolling updates are happening.

We make the assumption that key updates are only done after previous keys have been sync'd through all nodes. And the assumption that higher numbered keys are newer with care to handle the rollover case where a keyID == 15 is older than keyID == 1.

Note: I decided to make 15 the max key because we can't have more than two keys in use at the same time now assuming users follow the above requirement that only allows for changing keys after previous update is complete. So having large key ids doesn't seem necessary and by making the max key 15 we save 4 bits for some other use later.

Fixes: #7590

Update: I was hitting some instruction limits so simplified my code and did some refactoring on the bpf_lxc code. I may have gained a few more instructions than needed so I could look at removing some of the refactoring but overall I like it so left it for now. There is enough free instructions now I can even add a single printk or cilium_dbg if I want to debug some specific aspect not already covered by existing dbg stmts.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7611)
<!-- Reviewable:end -->

nebril commented on bpf/lib/l3.h:
```diff
@@ -135,4 +135,35 @@ static inline int __inline__ ipv4_local_delivery(struct __sk_buff *skb, int l3_o
+	 * will always be returned if either local or peer have the zero
+	 * key indicating no encryption.
+	 */
+	if (peer_key == 15)
```
Can you write somewhere where does `15` come from?

jrfastab replied:
Sure I'll put it as a default in linux_defaults, its a bit arbitrary but using 4 bits we can have 15 keys.

jrfastab committed:
cilium: transparent encryption, use correct keys during key rotation

When keys are being changed from previous key to new key its possible
that nodes will have different sets of keys installed in the datapath.
This is a result of either a rolling update and/or because secrets are
not propagated at the same time through all the nodes. The result is
a peer key ID (e.g. the one looked up in the ipcache) may represent
a key that has not yet been installed on the node. Trying to encrypt
with that key will result in dropped traffic because the key is
not known. So this patch keeps the state of the last key installed in
the datapath and uses that. The assumption key updates only happen
after all nodes have sync'd the prevous update.

This required implementing another map to hold the state.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
Revert "test: Disable flaky encapsulation encryption test"

After fix to correctly handle peer and local keys during key transitions
e.g. enabling/disabling encryption like this test. And reducing instruction
count to get this running on all kernels and LLVM7. Lets enable now.

This reverts commit 08e4dde73a327bf693422df76a67e99f5d5a8655.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: remove unecessary zero'ing of ip6 endpoint key

The structure is initialized at initialization no need to zero the
fields yet again. Apparently compiler does emit these instructions.

To verify change reduces complexity use Vagrant box provided with
Cilium and the following command,

#make -C bpf && ./test/bpf/check-complexity.sh

Before

Prog section '2/10' (tail_call IPV6_FROM_LXC) loaded (40)!
 - Instructions: 3746 (0 over limit)
processed 49004 insns

After

 Prog section '2/10' (tail_call IPV6_FROM_LXC) loaded (40)!
  - Instructions: 3743 (0 over limit)
 processed 48977 insns

Note, the max instruction limit is 4096 and we will be using the
remaining instructions for fixing tunnel keys when encryption keys
are changing.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: remove relax() calls to get more free insns

This paticular relax() call does little for complexity limits and I
need the extra instructions more.

Testing on Vagrant box provided with cilium/test which is using a
4.9 kernel.

Before:

  Prog section '2/10' (tail_call IPV6_FROM_LXC)
   - Instructions: 3743 (0 over limit)
  processed 48977 insns

After:

 Prog section '2/10' (tail_call IPV6_FROM_LXC) loaded (40)!
  - Instructions: 3735 (0 over limit)
 processed 49451 insns

For the win +8 instructions!!! =)

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: combine tunnel and non-tunnel cases into single branch

Refactoring tunnel and non-tunnel cases in encap.h reduces the insn
count a reasonable amount. It also helps identify in encap.h where
each call is from. We now have a _lxc encap call and a _netdev encap
call to handle calls from ingress of veth pair attached to pods and
the cilium_host running bpf_netdev.

Before:
Prog section '2/10' (tail_call IPV6_FROM_LXC) loaded (40)!
      - Instructions: 3735 (0 over limit)
     processed 49451 insnser:

After:
Prog section '2/10' (tail_call IPV6_FROM_LXC) loaded (40)!
 - Instructions: 3711 (0 over limit)
processed 48850 insns

Gain another 24 instructions.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

jrfastab committed:
cilium: fix dropping Health node IP updates

In my failing ginkgo test for cillium + vxlan + encryption I see Preflight
failure. After reviewing the logs I noticed that one node is emmitting an
event with only the Health IPs changing and in this case the peer is
complaining that it is already in the ipcache and therefore is not updating
its node map.

This results in a situation where the map has the old Health IPs and when
we run cilium-health we try to http/ping the old non-existing IP and as
expected get a timeout.

Running 'cilium node list' on each node shows that the two nodes disagree
on the Health IPs. It is also always exactly one of the two nodes in my
setup.

To fix this update the node map (but don't do datpath reconfigs) on all
nodes even if the ipcache has an entry for the node. We can skip the
datapath configuration to avoid churn because the datapath does not use
the health IPs.

Signed-off-by: John Fastabend <john.fastabend@gmail.com>

ianvernon reviewed (approved):
Approving for CI mainly - excited to see this test re-enabled!

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -1817,6 +1817,7 @@ func (d *Daemon) updateK8sNodeTunneling(k8sNodeOld, k8sNodeNew *types.Node) erro
 		Source: ipcache.FromKubernetes,
 	})
 	if !selfOwned {
+		d.nodeDiscovery.Manager.NodeSoftUpdated(*nodeNew)
```
Newbie question: how come the datapath does not have to be updated here? E.g., why use `NodeSoftUpdated` vs. `NodeUpdated`?

joestringer replied:
I believe that the `ipcache.IPIdentityCache.Upsert()` would already trigger such a datapath update. 

If I follow, the problem here is that we previously wouldn't issue any notification to the manager that the node has potentially changed, so then when a health IP shows up, this code wouldn't propagate that information through to the API (and then through to `cilium-health`, which would then ping the new IP instead of the old one)

joestringer commented on pkg/maps/encrypt/encrypt.go:
```diff
@@ -0,0 +1,102 @@
+
+// EncryptKey is the context ID for the encryption session
+type EncryptKey struct {
+	key uint32 `align:"ctx"`
```
nit: I'd expect the BPF side of this to show up in `bpf/bpf_alignchecker.c` for this align tag to mean something. That said, With a one-field struct I'm not sure whether there's a way for the alignment to get mixed up..

joestringer reviewed (approved):
Minor nit, though I don't see any likely case it'll cause trouble.
